For main repo

ThunderstormXX opened the pull request:
modules.py - Добавил обработку ребер с различными rho, mu
my_algs.py - Новые алгоритмы
algs.py - Немного изменил алгоритмы : возвращает время работы вторым аргументом ,также перенес туда все алгоритмы из my_algs.py
test.py - Скрипты для запуска экспериментов
compare_methods.py - Пример запуска экспериментов с использованием test.py
experiments_results - результаты экспериментов (там можно ничего не чекать - там данные)
TransportationNetworks - добавил все датасеты из bstabler

niquepolice commented on Readme.md:
```diff
@@ -46,6 +46,7 @@ Docker image  might be created on demand to simplify the installation process. W
 [SemyonSemenov](https://github.com/SemyonSemenov/mmo_tm) Sinkhorn GPU (pytorch) implemetation and experiments \[implementation merged\]
 
 
+
```
убрать

niquepolice commented on src/test.py:
```diff
@@ -0,0 +1,132 @@
+
```
зачем этот файл нужен?

niquepolice commented on Readme.md:
```diff
@@ -46,6 +46,7 @@ Docker image  might be created on demand to simplify the installation process. W
 [SemyonSemenov](https://github.com/SemyonSemenov/mmo_tm) Sinkhorn GPU (pytorch) implemetation and experiments \[implementation merged\]
```
Давай все эти лианы тащить не будем, свернём все изменения в один коммит. Предлагаю тебе просто завести ветку ровно на мастере основного репозитория, скопировать все изменённые файлы из этой ветки в неё и закоммитить одним коммитом. И описание коммита на английском надо

![image](https://github.com/niquepolice/mmo_tm/assets/27158260/30e6c632-42ab-4bd7-aba3-bbbb5a4a71be)

niquepolice replied:
чтобы просто вот так было 
![image](https://github.com/niquepolice/mmo_tm/assets/27158260/6cf5b1d4-b34e-4fe0-a02f-66ae708481c0)

niquepolice commented on all_models.ipynb:
```diff
@@ -2,7 +2,7 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
```
Удалить аутпуты из ноутбука

niquepolice commented on all_models.ipynb:
```diff
@@ -110,10 +127,32 @@
+    }
+   ],
    "source": [
     "times_e_ustm, flows_e_ustm, logs, optimal = ustm(beckmann_model, eps_abs, max_iter=1000, stop_by_crit=True)\n",
```
```suggestion
    "times_e_ustm, flows_e_ustm, logs, optimal = ustm(beckmann_model, eps_abs, max_iter=1000, stop_by_crit=False)\n",
```

niquepolice commented on all_models.ipynb:
```diff
@@ -122,28 +161,83 @@
+    }
+   ],
    "source": [
     "times_e_fw, flows_e_fw, logs, optimal = frank_wolfe(beckmann_model, eps_abs,  max_iter=1000, stop_by_crit=False)\n",
```
```suggestion
    "times_e_fw, flows_e_fw, logs, optimal = frank_wolfe(beckmann_model, eps_abs,  max_iter=4000, stop_by_crit=False)\n",
```

niquepolice commented on all_models.ipynb:
```diff
@@ -122,28 +161,83 @@
+    }
+   ],
+   "source": [
+    "times_e_nfw, flows_e_nfw, logs, optimal = N_conjugate_frank_wolfe(beckmann_model, eps_abs,  max_iter=1000, stop_by_crit=False, cnt_conjugates=3)\n",
```
```suggestion
    "times_e_nfw, flows_e_nfw, logs, optimal = N_conjugate_frank_wolfe(beckmann_model, eps_abs,  max_iter=4000, stop_by_crit=False, cnt_conjugates=3)\n",
```

niquepolice commented on all_models.ipynb:
```diff
@@ -122,28 +161,83 @@
     "plt.plot(dgap_ustm, c=\"C4\", label=\"USTM\")\n",
     "plt.plot(dgap_fw, c=\"C5\", label=\"FW\")\n",
+    "plt.plot(dgap_nfw, c=\"C3\", label=\"NFW\")\n",
     "plt.plot(np.ones(max(dgap_ustm.size, dgap_fw.size)) * eps_abs, label=\"eps_abs\")\n",
```
```suggestion
    "plt.plot(np.ones(max(dgap_ustm.size, dgap_fw.size)) * eps_abs, `--`, label=\"eps_abs\")\n",
```

niquepolice replied:
Предлагаю поставить eps=1e-4, увеличить число итераций у франкВульфов и точность обозначить пунктиром, тогда симпотнее выглядеть будет и будет видно, что NFW лучше работает
![image](https://github.com/niquepolice/mmo_tm/assets/27158260/9f22e247-827f-4c6c-a303-dc3a823772cd)

niquepolice replied:
теперь ваще красота

niquepolice commented on src/algs.py:
```diff
@@ -27,42 +28,210 @@ def frank_wolfe(
+        # if max_dual_func_val == dual_val  :
+        #     steps.append(k)
 
         # equal to FW gap if dual_val == max_dual_func_val
```
коммент не к этой строчке?

niquepolice commented on src/algs.py:
```diff
@@ -27,42 +28,210 @@ def frank_wolfe(
 
         dual_val = model.dual(times, flows)
         max_dual_func_val = max(max_dual_func_val, dual_val)
+        # print(dual_val ,max_dual_func_val)
```
убрать все ненужное

niquepolice commented on src/models.py:
```diff
@@ -71,7 +71,21 @@ def __init__(self, nx_graph: nx.DiGraph, correspondences: Correspondences):
+        fft, mu, rho, caps = get_graph_props(self.graph)
+        
+        # np.where(np.isinf(mu))
+        self.is_inf = np.where(rho == 0)
```
наверно можно убрать везде `np.where`, чтобы вместо массивов индексов работать с булевыми масками. От них отрицание брать можно просто через `~`, без всяких `np.isin`

niquepolice replied:
мб название на что-то более конкретное заменить?  Типа `centroid_connection` и добавить коммент, что это дуги с `rho=0`, используемые для привязки центроидов. Или назвать `mu_is_inf` или `rho_is_zero`. Ещё сейчас называется isinf, а проверка на 0)

niquepolice commented on src/models.py:
```diff
@@ -119,27 +133,53 @@ class BeckmannModel(TrafficModel):
         fft, mu, rho, caps = get_graph_props(self.graph)
+        
+        result = np.empty(len(mu))
+        result[self.is_inf] = fft[self.is_inf]*(1 + rho[self.is_inf])
```
`rho[self.is_inf] == 0`, как то странно его прибавлять тогда. Предлагаю с неймингами разобраться и ещё раз посмотреть

niquepolice commented on src/models.py:
```diff
@@ -119,27 +133,53 @@ class BeckmannModel(TrafficModel):
+
+        result[self.is_inf] = 0
+        # тут два случая : 1) rho != 0 and fft = 0 2) rho != 0 and fft != 0
+        result[self.fft_is_not_zero] = caps[self.fft_is_not_zero] * (dt[self.fft_is_not_zero] / (fft[self.fft_is_not_zero] * rho[self.fft_is_not_zero])) ** mu[self.fft_is_not_zero] * dt[self.fft_is_not_zero] / (1 + mu[self.fft_is_not_zero])
```
я бы завел переменную `mask=self.fft_is_not_zero`, чтобы формула чуть читаемее выглядела

ThunderstormXX committed:
Add NFW and experiments

Add  algorithm NFW and experiments. Cases of value hyperparameters of edge is reviewed.

niquepolice commented on all_models.ipynb:
```diff
@@ -58,17 +58,21 @@
     "# net_name = \"Anaheim_net\"\n",
     "# traffic_mat_name = \"Anaheim_trips\"\n",
     "    \n",
+    "#705 565 0 from 2522 вложено\n",
```
что значат комменты? код и гит на русском

niquepolice commented on src/algs.py:
```diff
@@ -27,42 +28,176 @@ def frank_wolfe(
+                gamma_list.pop(0)
+
+        if linesearch :
+            res = minimize_scalar( lambda y : model.primal(flows + y*dk ) , bounds = (0.0,1.0) , tol = 1e-12 )
```
у меня выбрасывается варнинг
`/home/demyan/.local/lib/python3.11/site-packages/scipy/optimize/_minimize.py:892: RuntimeWarning: Method 'bounded' does not support relative tolerance in x; defaulting to absolute tolerance.
  warn("Method 'bounded' does not support relative tolerance in x; "`

я так понимаю `tol` надо заменить на `xatol`

niquepolice commented on src/models.py:
```diff
@@ -119,27 +134,53 @@ class BeckmannModel(TrafficModel):
         fft, mu, rho, caps = get_graph_props(self.graph)
+        
+        result = np.empty(len(mu))
+        result[self.mu_mask] = fft[self.mu_mask]*(1 + rho[self.mu_mask])
```
почему не `all_cases` ?

niquepolice commented on src/algs.py:
```diff
@@ -27,42 +28,176 @@ def frank_wolfe(
 
-        # dgap_log.append(times @ (flows_averaged - flows))  # FW gap
+        if linesearch :
+            res = minimize_scalar( lambda y : model.primal(flows_averaged*(1-y) + y*flows ) , bounds = (0.0,1.0) , tol = 1e-12 )
```
надо бы ещё блеком пройтись, чтобы форматирование везде поправить

niquepolice commented on src/algs.py:
```diff
@@ -27,29 +28,42 @@ def frank_wolfe(
+            res = minimize_scalar(
+                lambda y: model.primal(flows_averaged * (1 - y) + y * flows),
+                bounds=(0.0, 1.0),
+                xatol = 1e-12
```
так не работает
```suggestion
                options={"xatol": 1e-12}
```

niquepolice commented on src/algs.py:
```diff
@@ -59,7 +73,144 @@ def frank_wolfe(
+
+        if linesearch:
+            res = minimize_scalar(
+                lambda y: model.primal(flows + y * dk), bounds=(0.0, 1.0), xatol = 1e-12            )
```
```suggestion
                lambda y: model.primal(flows + y * dk), bounds=(0.0, 1.0), 
                                options={"xatol": 1e-12})
```
